crates/astria-conductor/src/executor/queue.rs

```diff
@@ -0,0 +1,683 @@
+            let mut blocks: Vec<&SequencerBlockSubset> = tmp_blocks.values().collect();
+            blocks.sort();
+            for block in blocks {
+                output_blocks.push(block.clone());
+            }
+            let most_recent_height = output_blocks[output_blocks.len() - 1].height();
+            self.head_height = most_recent_height.increment();
+            self.remove_data_below_height(self.head_height);
+            return Some(output_blocks);
+        }
+        None
+    }
+
+    /// Check to see if the block is already present in the queue
+    fn is_block_present(&mut self, block: &SequencerBlockSubset) -> bool {
+        let block_hash = block.block_hash();
+        let height = block.height();
+
+        // check if the block is already present in the pending blocks
+        if let Some(pending_blocks) = self.pending_blocks.get(&height) {
+            if let Some(_block) = pending_blocks.get(&block_hash) {
+                return true;
+            }
+        }
+        // check if the block is already present in the soft blocks
+        if let Some(soft_block) = self.soft_blocks.get(&height) {
+            if soft_block.block_hash() == block_hash {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Check if there is another block in the pending blocks at a lower height
+    /// that points to this block as its parent.
+    fn is_block_a_parent(&mut self, block: SequencerBlockSubset) -> bool {
+        let block_hash = block.block_hash();
+        if let Some(child_blocks) = self.pending_blocks.get(&block.child_height()) {
+            let blocks = child_blocks.values();
+            for block in blocks {
+                if let Some(hash) = block.parent_hash() {
+                    if hash == block_hash {
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Safely insert a block into the pending blocks. This function will insert
+    /// the block into existing maps or create a new one if needed.
+    fn insert_to_pending_blocks(&mut self, block: SequencerBlockSubset) {
+        let height = block.height();
+        let block_hash = block.block_hash();
+
+        if let Some(pending_blocks) = self.pending_blocks.get_mut(&height) {
+            let mut new_map = pending_blocks.clone();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        } else {
+            let mut new_map = HashMap::new();
+            new_map.insert(block_hash, block);
+            self.pending_blocks.insert(height, new_map);
+        }
+        self.update_internal_state();
+    }
+
+    // remove all data in the queue below a given height. this does not remove
+    // data from the soft queue, only the pending queue, and updates the head height.
+    // TODO: add error handling
+    fn remove_data_below_height(&mut self, height: Height) {
```

Comment:
rn this is always called with `self.head_height` passed as param, I think we can just load `self.head_height` inside the scope of the function.